table/tables/index.go

```diff
@@ -124,6 +126,32 @@ func (c *index) getIndexKeyBuf(buf []byte, defaultCap int) []byte {
 	return make([]byte, 0, defaultCap)
 }
 
+// truncateIndexValuesIfNeeded truncate the index values that be created that use only the leading part of column values.
+func (c *index) truncateIndexValuesIfNeeded(indexedValues []types.Datum) []types.Datum {
+	for i := 0; i < len(indexedValues); i++ {
+		v := &indexedValues[i]
+		if v.Kind() == types.KindString || v.Kind() == types.KindBytes {
+			ic := c.idxInfo.Columns[i]
+			if ic.Tp.Charset == charset.CharsetUTF8 || ic.Tp.Charset == charset.CharsetUTF8MB4 {
+				val := v.GetBytes()
+				if ic.Length != types.UnspecifiedLength && utf8.RuneCount(val) > ic.Length {
```

Comment:
I think we can use utf8.RuneCountInString(v) instead, and thus eliminate the usage of val.

Reply:
But RuneCountInString need to convert bytes to string first, it's unworthy.

Reply:
I see.